terminating connection due to unexpected postmaster exit

Not sure if this happens often yet, but might be something we should try to improve with some connection validation (if it's not too costly) or retry.

```
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-2] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-1] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-1] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-2] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-4] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,399] [WARN] [io.r2dbc.postgresql.client.ReactorNettyClient] [] [] [reactor-tcp-epoll-3] [] - Error: SEVERITY_LOCALIZED=FATAL, SEVERITY_NON_LOCALIZED=FATAL, CODE=57P01, MESSAGE=terminating connection due to unexpected postmaster exit, FILE=be-secure.c, LINE=199, ROUTINE=secure_read  MDC: {}
[2021-11-22 16:43:33,499] [ERROR] [akka.projection.testing.ConfigurablePersistentActor$] [akka://appka/system/sharding/configurable/220/test-1637599365787-1162] [] [appka-akka.actor.default-dispatcher-12] [configurable|test-1637599365787-1162] - Supervisor StopSupervisor saw failure: Failed to persist event type [akka.persistence.journal.Tagged] with sequence number [2] for persistenceId [configurable|test-1637599365787-1162]  MDC: {persistencePhase=persist-evt, akkaAddress=akka://appka@10.8.4.4:2552, akkaSource=akka://appka/system/sharding/configurable/220/test-1637599365787-1162, sourceActorSystem=appka, persistenceId=configurable|test-1637599365787-1162}
akka.persistence.typed.internal.JournalFailureException: Failed to persist event type [akka.persistence.journal.Tagged] with sequence number [2] for persistenceId [configurable|test-1637599365787-1162]
	at akka.persistence.typed.internal.Running$PersistingEvents.onJournalResponse(Running.scala:745)
	at akka.persistence.typed.internal.Running$PersistingEvents.onMessage(Running.scala:647)
	at akka.persistence.typed.internal.Running$PersistingEvents.onMessage(Running.scala:632)
	at akka.actor.typed.scaladsl.AbstractBehavior.receive(AbstractBehavior.scala:84)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:57)
	at akka.persistence.typed.internal.EventSourcedBehaviorImpl$$anon$1.aroundReceive(EventSourcedBehaviorImpl.scala:212)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:85)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:57)
	at akka.actor.typed.internal.SimpleSupervisor.aroundReceive(Supervision.scala:131)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:85)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:57)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:87)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
	at akka.actor.typed.internal.InterceptorImpl$$anon$2.apply(InterceptorImpl.scala:57)
	at akka.actor.typed.internal.SimpleSupervisor.aroundReceive(Supervision.scala:131)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:85)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:274)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:230)
	at akka.actor.typed.internal.adapter.ActorAdapter.handleMessage(ActorAdapter.scala:131)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:107)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
	at akka.actor.ActorCell.invoke(ActorCell.scala:548)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: akka.pattern.CircuitBreaker$$anon$13: Circuit Breaker Timed out.
```

--- a/core/src/main/resources/reference.conf
+++ b/core/src/main/resources/reference.conf
@@ -99,12 +99,21 @@ akka.persistence.r2dbc {
       root-cert = ""
     }
 
+    # Initial pool size.
     initial-size = 5
+    # Maximum pool size.
     max-size = 20
+    # Maximum time to create a new connection.
     create-timeout = 3 seconds
+    # Maximum time to acquire connection from pool.
     acquire-timeout = 4 seconds
-    acquire-retry = 3
+    # Number of retries if the connection acquisition attempt fails.
+    # Use the same number as max-size in case all connections in the pool
+    # are invalid because database server was restarted.
+    acquire-retry = 20
 
+    # Maximum idle time of the connection in the pool.
+    # This value is used as an interval for background eviction of idle connections.
     max-idle-time = 30 minutes
 
     # Configures the statement cache size.
@@ -112,6 +121,11 @@ akka.persistence.r2dbc {
     # a positive value will configure a bounded cache with the passed size.
     # TODO: add more elaborated description explaining when and why to use a cache
     statement-cache-size = 100
+
+    # Validate the connection when acquired with this SQL.
+    # Enabling this has some performance overhead.
+    # A fast query for Postgres is "SELECT 1"
+    validation-query = ""
   }
 
   # If database timestamp is guaranteed to not move backwards for two subsequent

--- a/core/src/main/scala/akka/persistence/r2dbc/ConnectionFactoryProvider.scala
+++ b/core/src/main/scala/akka/persistence/r2dbc/ConnectionFactoryProvider.scala
@@ -95,13 +95,13 @@ class ConnectionFactoryProvider(system: ActorSystem[_]) extends Extension {
       .maxAcquireTime(JDuration.ofMillis(settings.acquireTimeout.toMillis))
       .acquireRetry(settings.acquireRetry)
       .maxIdleTime(JDuration.ofMillis(settings.maxIdleTime.toMillis))
-//      .validationQuery("SELECT 1") // FIXME issue #157
-      .build()
 
-    val pool = new ConnectionPool(poolConfiguration)
+    if (settings.validationQuery.nonEmpty)
+      poolConfiguration.validationQuery(settings.validationQuery)
+
+    val pool = new ConnectionPool(poolConfiguration.build())
 
     // eagerly create initialSize connections
-
     pool.warmup().asFutureDone() // don't wait for it
 
     pool

--- a/core/src/main/scala/akka/persistence/r2dbc/R2dbcSettings.scala
+++ b/core/src/main/scala/akka/persistence/r2dbc/R2dbcSettings.scala
@@ -78,11 +78,13 @@ final class ConnectionFactorySettings(config: Config) {
 
   val initialSize: Int = config.getInt("initial-size")
   val maxSize: Int = config.getInt("max-size")
+  val maxIdleTime: FiniteDuration = config.getDuration("max-idle-time").asScala
+
   val createTimeout: FiniteDuration = config.getDuration("create-timeout").asScala
   val acquireTimeout: FiniteDuration = config.getDuration("acquire-timeout").asScala
   val acquireRetry: Int = config.getInt("acquire-retry")
 
-  val maxIdleTime: FiniteDuration = config.getDuration("max-idle-time").asScala
+  val validationQuery: String = config.getString("validation-query")
 
   val statementCacheSize: Int = config.getInt("statement-cache-size")
 }

